src/cmake/Helpers.cmake

```diff
@@ -0,0 +1,186 @@
+  endif()
+
+  add_library(${interface_name} INTERFACE)
+
+  target_link_libraries(${interface_name}
+    INTERFACE ${LIBRARY_LINKS}
+  )
+
+  install(TARGETS ${interface_name}
+    EXPORT aliceVision-targets
+  )
+
+  set(LIBRARY_NAME_INTERFACE "${interface_name}_interface")
+  add_custom_target(${LIBRARY_NAME_INTERFACE} SOURCES ${LIBRARY_SOURCES})
+
+  set_property(TARGET ${LIBRARY_NAME_INTERFACE}
+    PROPERTY FOLDER "AliceVision"
+  )
+endfunction()
+
+# Add software function
+function(alicevision_add_software software_name)
+  set(options "")
+  set(singleValues FOLDER)
+  set(multipleValues SOURCE LINKS INCLUDE_DIRS)
+
+  cmake_parse_arguments(SOFTWARE "${options}" "${singleValues}" "${multipleValues}" ${ARGN})
+
+  if(NOT software_name)
+    message(FATAL_ERROR "You must provide the software name in 'alicevision_add_software'")
+  endif()
+
+  if(NOT SOFTWARE_SOURCE)
+    message(FATAL_ERROR "You must provide the software SOURCE in 'alicevision_add_software'")
+  endif()
+
+  if(NOT SOFTWARE_FOLDER)
+    message(FATAL_ERROR "You must provide the software FOLDER in 'alicevision_add_software'")
+  endif()
+
+  add_executable(${software_name} ${SOFTWARE_SOURCE})
+
+  target_link_libraries(${software_name}
+    PUBLIC ${SOFTWARE_LINKS}
+  )
+
+  target_include_directories(${software_name}
+    PUBLIC ${SOFTWARE_INCLUDE_DIRS}
+  )
+
+  set_property(TARGET ${software_name}
+    PROPERTY FOLDER ${SOFTWARE_FOLDER}
+  )
+
+  # set_target_properties(${software_name}
+  #   PROPERTIES SOVERSION ${ALICEVISION_VERSION_MAJOR}
+  #   VERSION "${ALICEVISION_VERSION_MAJOR}.${ALICEVISION_VERSION_MINOR}"
+  # )
+
+  install(TARGETS ${software_name}
+    DESTINATION bin/
+  )
+endfunction()
+
+# Add test function
+function(alicevision_add_test test_file)
+  set(options "")
+  set(singleValues NAME)
+  set(multipleValues LINKS INCLUDE_DIRS)
+
+  cmake_parse_arguments(TEST "${options}" "${singleValues}" "${multipleValues}" ${ARGN})
+
+  if(NOT test_file)
+    message(FATAL_ERROR "You must provide the test file in 'alicevision_add_test'")
+  endif()
+
+  if(NOT TEST_NAME)
+    message(FATAL_ERROR "You must provide the NAME in 'alicevision_add_test'")
+  endif()
+
+  if(NOT ALICEVISION_BUILD_TESTS)
+    return()
+  endif()
+
+  set(TEST_EXECUTABLE_NAME "aliceVision_test_${TEST_NAME}")
+
+  add_executable(${TEST_EXECUTABLE_NAME} ${test_file})
+
+  target_link_libraries(${TEST_EXECUTABLE_NAME}
+    PUBLIC ${TEST_LINKS}
+           ${ALICEVISION_LIBRARY_DEPENDENCIES}
+           ${Boost_LIBRARIES}
+  )
+
+  target_include_directories(${TEST_EXECUTABLE_NAME}
+    PUBLIC ${TEST_INCLUDE_DIRS}
+           ${Boost_INCLUDE_DIRS}
+  )
+
+  set_property(TARGET ${TEST_EXECUTABLE_NAME}
+    PROPERTY FOLDER Test
+  )
+
+  add_test(NAME ${TEST_EXECUTABLE_NAME}
+           WORKING_DIRECTORY ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}
+           COMMAND $<TARGET_FILE:${TEST_EXECUTABLE_NAME}> --catch_system_error=yes --log_level=all
+  )
```

Comment:
Would be good to declare the dependency between the test execution and the executable with `add_dependencies(<target> [<target-dependency>]...)`:
`add_dependencies("aliceVision_testexec_${TEST_NAME}" ${TEST_EXECUTABLE_NAME})`

Reply:
Sorry but I don't understand.
I can't do like your example. In `add_test` NAME is not a target.
```
  CMake Error at src/cmake/Helpers.cmake:193 (add_dependencies):
  Cannot add target-level dependencies to non-existent target
  "aliceVision_testexec_foo".

  The add_dependencies works for top-level logical targets created by the
  add_executable, add_library, or add_custom_target commands.  If you want to
  add file-level dependencies see the DEPENDS option of the add_custom_target
  and add_custom_command commands.
```
What did I misunderstand ?